### PR TITLE
Fix initialization bug

### PR DESCRIPTION
**Pull request by AlexanderWells-diamond**
During startup of the application, it attempts to connect to the "Prologix" controller, plugged in via USB port. This is the `connection` object in the `Worker` class. It starts uninitialized, and only gains a value when the "Prologix" controller is found (by trying all possible connected USB devices until something responds appropriately). 

It is possible for users to make changes in the GUI before the equipment is found or, more commonly, before the Prologix controller is plugged in. This triggers an attempt to write the new configuration to the controller.This caused exceptions that were causing the GUI to crash.

Now we explicitly guard the connection object, and raise a specific exception if it is not initialized. This allows the GUI to continue operating.

Note that this can cause an issue where configuration is changed on the GUI, but has not been written to the controller. For example, if all configuration items are set in the GUI before the Prologix controller is connected, and then it is connected, pressing the "start" button may cause logging to begin with unexpected configuration. I'm not sure how much time it is worth committing to fixing this  - it's probably a separate issue anyway.

### CODE REVIEW

File: src/psc_datalogger/gui.py [hunk 1]
```diff
@@ -129,12 +129,19 @@ def handle_start_stop(self, checked):
...
-                i.get_temperature_checked(),
-            )
+            try:
+                self.connection_manager.set_instrument(
+                    i.instrument_number,
+                    i.isChecked(),
+                    i.get_address(),
+                    i.get_temperature_checked(),
+                )
+            except ValueError:
```
Comment (coretl) [inline src/psc_datalogger/gui.py:139]: I can't see what would raise ValueError in the code?

**Review by coretl — COMMENTED** [review]
Apart from the ValueError it all looks sane, might be nice to surface the error on the GUI rather than just a logging.exception?

File: src/psc_datalogger/gui.py (hunk 1 above)
Comment (AlexanderWells-diamond) [reply src/psc_datalogger/gui.py:139]: It's a little convoluted - `Worker.set_instrument` will see an empty string for `gpib_address` when the "Enable" tickbox is first ticked for an instrument (as the address field has never had a chance to be filled). This sets the `address` variable to `-1`, which is then passed into `Worker._init_instrument`, which sees that `-1` and raises the `ValueError`

File: src/psc_datalogger/connection.py [hunk 2]
```diff
@@ -492,3 +509,41 @@ def _exit(self) -> None:
...
+        """Wrapper for accessing connection.bytes_in_buffer"""
+        self._connection_check()
+        return self.connection.bytes_in_buffer
+
+    def _connection_read(self) -> str:
+        """Wrapper for all connection read commands."""
+        self._connection_check()
+        return self.connection.read()
+
+    def _connection_timeout(self, timeout: int) -> None:
```
Comment (MJGaughran) [inline src/psc_datalogger/connection.py]: I would use a verb for this, e.g. set_connection_timeout. At the moment this uses a noun like 'connection_bytes_in_buffer', except it's actually a setter and behaves differently.

File: src/psc_datalogger/connection.py [hunk 3]
```diff
@@ -492,3 +509,41 @@ def _exit(self) -> None:
...
+        Mostly required to guard against potential checks before the Prologix
+        controller has connected (or maybe there isn't even one plugged in yet!)"""
+        if not hasattr(self, "connection"):
+            logging.warning(
+                "Attempted to write to connection when it was not initialized"
+            )
+            raise ConnectionNotInitialized()
+
+    def _connection_write(self, command: str) -> None:
+        """Wrapper for all connection write commands."""
```
Comment (MJGaughran) [inline src/psc_datalogger/connection.py]: The fact that it is a wrapper isn't that relevant as a description. You're checking the connection before attempting a write operation.

Also, the fact that these methods are all prepended with 'connection' suggests that this abstraction would make sense as a separate class etc, but this program is small enough that it isn't too important.

File: src/psc_datalogger/gui.py [hunk 4]
```diff
@@ -129,12 +129,19 @@ def handle_start_stop(self, checked):
...
+                # Expected when first activating an instrument; the Address field
+                # will be empty
+                logging.exception(
+                    f"ValueError when initializing instrument {i.instrument_number}"
```
Comment (MJGaughran) [inline src/psc_datalogger/gui.py]: This is an expected error, which is always captured and handled appropriately, but the Exception message here is not very useful.

Also, logging.exception will flag up this message as an error, but above you state that it is just the normal result when first activating an instrument? This should be INFO level or below.

**Comment by MJGaughran** [comment]
The fact that GUI changes aren't necessarily applied to the controller when pressing start does seem to be an issue to me. Also, does the screen update or not when the controller is plugged in, to reflect the current settings?

If the time to upload config is quite low, I would consider only applying configuration when you tell the device to start, rather than having to care about whether or not the device is connected and trying to sync the data etc.

File: src/psc_datalogger/connection.py (hunk 2 above)
Comment (AlexanderWells-diamond) [reply src/psc_datalogger/connection.py]: Updated in 12288e0aab432a492f497cddbf445f9c2dd1fce5

File: src/psc_datalogger/connection.py (hunk 3 above)
Comment (AlexanderWells-diamond) [reply src/psc_datalogger/connection.py]: Updated in 12288e0aab432a492f497cddbf445f9c2dd1fce5

File: src/psc_datalogger/gui.py (hunk 4 above)
Comment (AlexanderWells-diamond) [reply src/psc_datalogger/gui.py]: I've ended up downgrading this to a `warning`, but moving it down to `info` would mean it wouldn't appear on the default logging level configuration. It's also possible that this error could be coming from something else and not just that expected initialization error; hence why I do want it to still be visible.

Changed in 12288e0aab432a492f497cddbf445f9c2dd1fce5

**Comment by AlexanderWells-diamond** [comment]
Currently the only update that happens when a Prologix controller is plugged in is that the status bar reads "Serial Connection open". This replaces an error message if the Prologix controller was not found during initialization.

I've made the requested changes. I think it's a separate piece of work to restructure this to make it simpler - probably by just removing all of the writing to the controller until after "Start" is pressed. There's no particular advantage to having all data written as soon as it changes in the GUI.

**Pull request merged**